### PR TITLE
Add dial option to set balancer

### PR DESCRIPTION
**Pull request by menghanl**
WithBalancerName dial option specifies the name of the balancer to be used by the ClientConn. Service config updates can NOT override the balancer option.

#1694

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc-go/1697)

### CODE REVIEW

File: clientconn.go [hunk 1]
```diff
@@ -188,7 +190,7 @@ func WithDecompressor(dc Decompressor) DialOption {
 
 // WithBalancer returns a DialOption which sets a load balancer with the v1 API.
 // Name resolver will be ignored if this DialOption is specified.
-// Deprecated: use the new balancer APIs in balancer package instead.
+// Deprecated: use the new balancer APIs in balancer package and WithBalancerName.
```
Comment (dfawley) [inline clientconn.go]: Nit: add a blank comment line above this one so it's its own paragraph.

File: clientconn.go [hunk 2]
```diff
@@ -206,6 +208,15 @@ func WithBalancerBuilder(b balancer.Builder) DialOption {
 	}
 }
 
+// WithBalancerName sets the balancer that the ClientConn will be initialized
+// with. The balancer can not be overridden by balancer option specified by
```
Comment (dfawley) [inline clientconn.go]: Nit: "cannot" or "will not".

Also, maybe something like: "Balancers are registered by name using the balancer package" before this sentence?

File: clientconn.go [hunk 3]
```diff
@@ -210,9 +210,9 @@ func WithBalancer(b Balancer) DialOption {
 	}
 }
 
-// WithBalancerBuilder is for testing only. Users using custom balancers should
+// withBalancerBuilder is for testing only. Users using custom balancers should
 // register their balancer and use service config to choose the balancer to use.
-func WithBalancerBuilder(b balancer.Builder) DialOption {
+func withBalancerBuilder(b balancer.Builder) DialOption {
```
Comment (dfawley) [inline clientconn.go]: Please delete, then LGTM.

**Comment by predmond** [comment]
Is this getting merged? I could use this feature..

File: clientconn.go [hunk 4]
```diff
@@ -670,9 +678,9 @@ func (cc *ClientConn) handleResolvedAddrs(addrs []resolver.Address, err error) {
 
 	cc.curAddresses = addrs
```
Comment (menghanl) [inline clientconn.go]: @dfawley I updated the following `if else` while doing a rebase. Please take another look. Thanks!

**Pull request merged**